agent-control/src/on_host/file_store.rs

```diff
@@ -0,0 +1,775 @@
+use std::{
+    io::{self, Error, ErrorKind},
+    path::{Path, PathBuf},
+    sync::RwLock,
+};
+
+use fs::{
+    LocalFile,
+    directory_manager::{DirectoryManager, DirectoryManagerFs},
+    file_reader::{FileReader, FileReaderError},
+    writer_file::FileWriter,
+};
+use serde::{Serialize, de::DeserializeOwned};
+use tracing::{debug, error, trace};
+
+use crate::{
+    agent_control::{
+        agent_id::AgentID,
+        defaults::{FOLDER_NAME_FLEET_DATA, FOLDER_NAME_LOCAL_DATA},
+    },
+    opamp::data_store::{OpAMPDataStore, StoreKey},
+};
+
+pub struct FileStore<F, D>
+where
+    D: DirectoryManager,
+    F: FileWriter + FileReader,
+{
+    directory_manager: D,
+    file_rw: F,
+    remote_dir: RwLock<RemoteDir>,
+    local_dir: LocalDir,
+}
+
+pub struct LocalDir(PathBuf);
+
+impl LocalDir {
+    pub fn get_local_file_path(&self, agent_id: &AgentID, key: &StoreKey) -> PathBuf {
+        self.0
+            .join(FOLDER_NAME_LOCAL_DATA)
+            .join(agent_id)
+            .join(build_config_name(key))
+    }
+}
+
+pub struct RemoteDir(PathBuf);
+
+impl RemoteDir {
+    pub fn get_remote_file_path(&self, agent_id: &AgentID, key: &StoreKey) -> PathBuf {
+        self.0
+            .join(FOLDER_NAME_FLEET_DATA)
+            .join(agent_id)
+            .join(build_config_name(key))
+    }
+}
```

Comment:
Nit: I like the "aliases", but the usage of their functions kind of repeat the name.

```
local_dir.get_local_file_path
remote_dir.get_remote_file_path
```

Should we rename them to `get_file_path`? Or maybe have a single `DirBuilder` with both functions? I don't know.